### PR TITLE
fips: Use salt >= 16 bytes in PBKDF2 selftest

### PR DESCRIPTION
**Pull request by neverpanic**
NIST SP 800-132 [\[1\]][1] section 5.1 says "[t]he length of the randomly-generated portion of the salt shall be at least 128 bits", which implies that the salt for PBKDF2 must be at least 16 bytes long (see also Appendix A.2.1).

The FIPS 140-3 IG [\[2\]][2] section 10.3.A requires that "the lengths and the properties of the Password and Salt parameters, as well as the desired length of the Master Key used in a CAST shall be among those supported by the module in the approved mode."

As a consequence, the salt length in the self test must be at least 16 bytes long for FIPS 140-3 compliance. Switch the self test to use the only test vector from RFC 6070 that uses salt that is long enough to fulfil this requirement. Since RFC 6070 does not provide expected results for PBKDF2 with HMAC-SHA256, use the output from [\[3\]][3], which was generated with python cryptography, which was tested against the RFC 6070 vectors with HMAC-SHA1.

 [1]: https://doi.org/10.6028/NIST.SP.800-132
 [2]: https://csrc.nist.gov/CSRC/media/Projects/cryptographic-module-validation-program/documents/fips%20140-3/FIPS%20140-3%20IG.pdf
 [3]: https://github.com/brycx/Test-Vector-Generation/blob/master/PBKDF2/pbkdf2-hmac-sha2-test-vectors.md

##### Checklist

- [ ] ~documentation is added or updated~
- [x] tests are added or updated

### CODE REVIEW

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
pycrypto uses OpenSSL underneath doesn't it?

What about repurposing one of the test vectors from `test/recipes/30-test_evp_data/evpkdf_pbkdf2.txt`?

**Comment by neverpanic** [comment]
> pycrypto uses OpenSSL underneath doesn't it?
> 
> What about repurposing one of the test vectors from `test/recipes/30-test_evp_data/evpkdf_pbkdf2.txt`?

Any reason why those would be more trustworthy?

In fact, I'm using exactly the vector that's in that file at line 95, except with truncated output: https://github.com/openssl/openssl/blob/master/test/recipes/30-test_evp_data/evpkdf_pbkdf2.txt#L95-L100

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
I'm not sure of their provenance but if they weren't generated using OpenSSL, I'd consider them a better test than using vectors generated by the code under test.

File: providers/fips/self_test_data.inc [hunk 1]
```diff
@@ -354,19 +354,27 @@ static const ST_KAT_PARAM x963kdf_params[] = {
 };
 
 static const char pbkdf2_digest[] = "SHA256";
+/* Input parameters from RFC 6070, vector 5 (because it is the only one with
+ * a salt >= 16 bytes, which NIST SP 800-132 section 5.1 requires). The
+ * expected output is taken from
+ *   https://github.com/brycx/Test-Vector-Generation/blob/master/PBKDF2/pbkdf2-hmac-sha2-test-vectors.md,
+ * which ran these test vectors with SHA-256. */
```
Comment (t8m) [inline providers/fips/self_test_data.inc:361]: Please fix the comment formatting - beginning and end of comment should be on a separate line.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by slontis** [comment]
3.0 also?

**Comment by paulidale** [comment]
This isn't required for 3.0 but should be harmless to include (famous last words).

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale @beldmit the comment formatting was fixed, please reapprove.

I assume there is no need to restart the 24 h wait for this style nit fix.

**Comment by beldmit** [comment]
I confirm reapproval

**Comment by t8m** [comment]
Merged to master and 3.1 branches. Thank you for your contribution.

**Pull request closed without merging**